### DISCUSSION TITLE
diagnosing/debugging erroneous errors in translation units

### DISCUSSION THREAD

**Discussion by mellery451** [clangd/clangd#2039] [Q&A]
My ask is a little broad, but I'm wondering what are some techniques for diagnosing clearly wrong errors being reported by clangd. In my specific case, I have loads of errors about some types not found. these types happen to be aliases of the form:

     using fooname = SomeTemplate<int32_t, 3, 4>;  

These are declared in a header file many includes deep, but I know it's included indirectly. Doesn't matter because I've even tried including this header file directly as the first include and the errors persists. Also, FWIW, this is a CUDA source file I'm dealing with, so perhaps there is additional subtle complexity there.

So I'm just curious about what kind of extra diagnostics or flags I can enable to try to track this down. Can I request that clangd emit a preprocessed file for instance just to confirm that it at least sees the text of my using declaration? I've tried verbose logging and all I see are the same errors about "unknown type name 'fooname'" which I also get in my list of errors from clangd via the client.

My project I huge and complex, so distilling this down to something shareable is not possible. If I knew of the specific pattern that causes issues, I could probably create a small repro - but this is exactly what I'm trying to figure out in the first place. Any advice appreciated.

**Comment by HighCommander4** [comment]
At a high level, the first diagnostic step I'd suggest is to look at what compile command clangd is using to process your source file, try running an equivalent command on the command line, and see whether you get the same errors there.

A bit more concretely, if you look at [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs), they will contain a line of the form `ASTWorker building file <path>` followed by the compile command clangd is using.

You can turn this command into something you can run on the command line with a couple of minor modifications:
 * remove the `-resource-dir` argument
 * replace the argv[0] with a path to a `clang` compiler matching your clangd version (and ideally from the same source / package)

In my experience, most of the time the errors that you see in clangd will occur on the command line as well, at which point you've simplified the issue to "get this file to compile with clang".

Most of the time it's a matter of getting the flags right. Once you figure that out, you can get clangd to use the needed flags (either by arranging for `compile_commands.json` to contain them, or making tweaks using the [clangd config file](https://clangd.llvm.org/config.html)) and clangd should start working properly as well.

The situation where clangd gives you errors that you don't see when running the command on the command line, is less common. This may indicate a clangd bug, and you'd probably need to reduce the test case and file it.

**Comment by mellery451** [comment]
I'm using nvcc for the real building...but not all of those flags are supported I believe. Just so I'm clear, are you suggesting that I try to get my compilation working with clang++ (for cuda files) or just verify that my nvcc command works standalone ?

**Reply by HighCommander4** [reply]
The former. Clangd is a clang-based tool, so being able to compile (or at least have the frontend part of the compiler process, as with `-fsyntax-only`) the code with clang is a requirement for clangd to work properly.
